EKF/EKFGSF_yaw.cpp

```diff
@@ -0,0 +1,661 @@
+	} else {
+		// skip this fusion step because calculation is badly conditioned
+		return;
+	}
+
+	// calculate Kalman gain K  nd covariance matrix P
+	// autocode from https://github.com/priseborough/3_state_filter/blob/flightLogReplay-wip/calcK.txt
+	// and https://github.com/priseborough/3_state_filter/blob/flightLogReplay-wip/calcPmat.txt
+	const float t2 = P00*velObsVar;
+ 	const float t3 = P11*velObsVar;
+	const float t4 = velObsVar*velObsVar;
+	const float t5 = P00*P11;
+	const float t9 = P01*P10;
+	const float t6 = t2+t3+t4+t5-t9;
+	float t7;
+	if (fabsf(t6) > 1e-6f) {
+		t7 = 1.0f/t6;
+	} else {
+		// skip this fusion step
+		return;
+	}
+	const float t8 = P11+velObsVar;
+	const float t10 = P00+velObsVar;
+	float K[3][2];
+
+ 	K[0][0] = -P01*P10*t7+P00*t7*t8;
+	K[0][1] = -P00*P01*t7+P01*t7*t10;
+	K[1][0] = -P10*P11*t7+P10*t7*t8;
+	K[1][1] = -P01*P10*t7+P11*t7*t10;
+	K[2][0] = -P10*P21*t7+P20*t7*t8;
+	K[2][1] = -P01*P20*t7+P21*t7*t10;
+
+	const float t11 = P00*P01*t7;
+	const float t15 = P01*t7*t10;
+	const float t12 = t11-t15;
+	const float t13 = P01*P10*t7;
+	const float t16 = P00*t7*t8;
+	const float t14 = t13-t16;
+	const float t17 = t8*t12;
+	const float t18 = P01*t14;
+	const float t19 = t17+t18;
+	const float t20 = t10*t14;
+	const float t21 = P10*t12;
+	const float t22 = t20+t21;
+	const float t27 = P11*t7*t10;
+	const float t23 = t13-t27;
+	const float t24 = P10*P11*t7;
+	const float t26 = P10*t7*t8;
+	const float t25 = t24-t26;
+	const float t28 = t8*t23;
+	const float t29 = P01*t25;
+	const float t30 = t28+t29;
+	const float t31 = t10*t25;
+	const float t32 = P10*t23;
+	const float t33 = t31+t32;
+	const float t34 = P01*P20*t7;
+	const float t38 = P21*t7*t10;
+	const float t35 = t34-t38;
+	const float t36 = P10*P21*t7;
+	const float t39 = P20*t7*t8;
+	const float t37 = t36-t39;
+	const float t40 = t8*t35;
+	const float t41 = P01*t37;
+	const float t42 = t40+t41;
+	const float t43 = t10*t37;
+	const float t44 = P10*t35;
+	const float t45 = t43+t44;
+
+	const float min_var = 1e-6f;
+	_ekf_gsf[model_index].P(0,0) = fmaxf(P00-t12*t19-t14*t22 , min_var);
+	_ekf_gsf[model_index].P(0,1) = P01-t19*t23-t22*t25;
+	_ekf_gsf[model_index].P(0,2) = P02-t19*t35-t22*t37;
+	_ekf_gsf[model_index].P(1,0) = P10-t12*t30-t14*t33;
+	_ekf_gsf[model_index].P(1,1) = fmaxf(P11-t23*t30-t25*t33 , min_var);
+	_ekf_gsf[model_index].P(1,2) = P12-t30*t35-t33*t37;
+	_ekf_gsf[model_index].P(2,0) = P20-t12*t42-t14*t45;
+	_ekf_gsf[model_index].P(2,1) = P21-t23*t42-t25*t45;
+	_ekf_gsf[model_index].P(2,2) = fmaxf(P22-t35*t42-t37*t45 , min_var);
+
+	// force symmetry
+	_ekf_gsf[model_index].P.makeBlockSymmetric<3>(0);
+
+	float oldYaw = _ekf_gsf[model_index].X[2];
+	for (uint8_t obs_index = 0; obs_index < 2; obs_index++) {
+		// apply the state corrections including the compression scale factor
+		for (unsigned row = 0; row < 3; row++) {
+			_ekf_gsf[model_index].X[row] -= K[row][obs_index] * _ekf_gsf[model_index].innov(obs_index) * innov_comp_scale_factor;
+		}
+	}
```

Comment:
If you make K an matrix::Matrix<float, 3, 2> or (2,3) you can replace this with a single matrix multiplication.

Reply:
Done